### PR TITLE
enhance auto device map and support XPU 

### PR DESCRIPTION
**Pull request by xin3he**
#### Related issue: 
- https://github.com/intel/torch-xpu-ops/issues/2232
- https://github.com/intel/auto-round/issues/917
- https://github.com/intel/auto-round/issues/712
- https://github.com/intel/auto-round/issues/557

#### Env: 
4 cards 24G Intel GPU 

## Llama 8B MXFP4 (w/o torch.compie)
`auto-round --model meta-llama/Llama-3.1-8B-Instruct  --scheme MXFP4 --device 0,1,2,3 --low_gpu_mem_usage`
#### Auto device map result:
{'self_attn.q_proj': 'xpu:2', 'self_attn.k_proj': 'xpu:2', 'self_attn.v_proj': 'xpu:1', 'self_attn.o_proj': 'xpu:1', 'mlp.gate_proj': 'xpu:1', 'mlp.up_proj': 'xpu:2', 'mlp.down_proj': 'xpu:3'}
#### card_idx: max_memory_used
0: 19G, 1: 16G; 2: 13G; 3: 21G 

## Llama 8B W4A16 (w/o torch.compie)
`auto-round --model meta-llama/Llama-3.1-8B-Instruct  --scheme W4A16 --device 0,1 --low_gpu_mem_usage`
#### Auto device map result:
{'self_attn.q_proj': 'xpu:1', 'self_attn.k_proj': 'xpu:0', 'self_attn.v_proj': 'xpu:0', 'self_attn.o_proj': 'xpu:1', 'mlp.gate_proj': 'xpu:1', 'mlp.up_proj': 'xpu:1', 'mlp.down_proj': 'xpu:1'}
#### card_idx: max_memory_used
0: 19G, 1: 8G

#### `estimate_tuning_block_mem` function output:
- param_memory considers input activation if `enable_act_quant=True`
- output_memory = pick_samples * seq_len * out_features * element_size
- all memory consider grad by multiply 2.
```
layer_memory_dict = {
    'q_proj': {'param_memory': 0.018, 'output_memory': 0.024},
    'k_proj': {'param_memory': 0.018, 'output_memory': 0.024}, 
    'v_proj': {'param_memory': 0.018, 'output_memory': 0.024},
    'o_proj': {'param_memory': 0.018, 'output_memory': 0.008}
}  
input_output_memory = 4  # GB
additional_memory = 1  # GB (or 19 for XPU)
```

#### card_0_left_memory 
`card_0_left_memory = device_0_memory - input_output_memory - additional_memory - total_block_output_memory`

#### `_allocate_layers_to_devices` function:
```
Example:
    Input:
        device_memory = {"cuda:0": 30.0, "cuda:1": 40.0, "cuda:2": 40.0}
        layer_memory_dict = {
            "q_proj": {"param_memory": 4.0}, "k_proj": {"param_memory": 1.0},
            "v_proj": {"param_memory": 1.0}, "o_proj": {"param_memory": 4.0},
            "gate_proj": {"param_memory": 11.0}, "up_proj": {"param_memory": 11.0},
            "down_proj": {"param_memory": 11.0}
        }
        mem_per_param = 2.0

    Result (allocation order by size):
        1. gate_proj (22GB) -> cuda:2 (largest, prefer last device)
        2. up_proj (22GB) -> cuda:1 (2nd largest, prefer 2nd last device)
        3. down_proj (22GB) -> cuda:0 (3rd largest, cuda:0 has 30GB available)
        4. q_proj (8GB) -> cuda:2 (neighbor of gate_proj, continuity bonus)
        5. o_proj (8GB) -> cuda:2 (neighbor of q_proj, continuity bonus)
        6. k_proj (2GB) -> cuda:1 (neighbor of q_proj via original order)
        7. v_proj (2GB) -> cuda:1 (neighbor of k_proj, continuity bonus)
```

### CODE REVIEW

**Comment by xin3he** [comment]
FYI. 70B is still OOM with 4 cards

File: auto_round/utils/device.py [hunk 1]
```diff
@@ -458,34 +458,71 @@ def check_memory_availability(device, inputs, weight, org_seqlen, org_bs):
...
+                # If enable_act_quant, add input tensor memory to param_memory
+                if enable_act_quant:
+                    input_size = pick_samples * seq_len * in_features * element_size
+                    input_memory_gb = input_size / 1024**3
+                    param_memory_gb += input_memory_gb
+            else:
+                output_memory_gb = 0.0
+
+            # memory * 2, because it contains grad tensor.
+            layer_memory_dict[layer_name] = {"param_memory": param_memory_gb * 2, "output_memory": output_memory_gb * 2}
```
Comment (wenhuach21) [inline auto_round/utils/device.py]: tuning v is 2X, and min max scale could be safely ignored

File: auto_round/utils/device.py [hunk 2]
```diff
@@ -458,34 +458,71 @@ def check_memory_availability(device, inputs, weight, org_seqlen, org_bs):
...
 
+    layer_memory_dict = {}
     total_param_mem = 0
+
+    # Calculate batch_size and sequence_length from input_ids for output memory estimation
+    seq_len = input_ids[0].shape[1] if input_ids and len(input_ids[0].shape) >= 2 else 1
+    element_size = input_ids[0].element_size() if input_ids else 2  # Default to 2 bytes (fp16/bf16)
+
     for name, module in block.named_modules():
         if check_to_quantized(module):
```
Comment (wenhuach21) [inline auto_round/utils/device.py]: For non-quantized layers/norm layer/swiglu, both weights and activations also need to be included.
In addition, attention activations are assumed to incur a memory cost proportional to sequence_length².

File: auto_round/utils/device.py [hunk 3]
```diff
@@ -458,34 +458,71 @@ def check_memory_availability(device, inputs, weight, org_seqlen, org_bs):
     return False, seqlen, bs
 
 
-def estimate_tuning_block_mem(block: torch.nn.Module, input_ids: list[torch.Tensor]) -> tuple[float, float]:
+def estimate_tuning_block_mem(
+    block: torch.nn.Module, input_ids: list[torch.Tensor], pick_samples: int
```
Comment (wenhuach21) [inline auto_round/utils/device.py]: pick_samples->batch_size?

File: auto_round/utils/device.py [hunk 4]
```diff
@@ -628,8 +665,107 @@ def set_non_auto_device_map(
...
+        # Phase 1: Direct assign largest layers to higher-index devices first
+        if layer_idx < num_devices - 1:
+            return cuda_xpu_devices[-(layer_idx + 1)]
+
+        # Phase 2: Choose device with best score (memory + continuity)
+        best_device = None
+        best_score = float("-inf")
+        current_layer_order = layer_order[layer_name]
+
+        for device in cuda_xpu_devices:
```
Comment (wenhuach21) [inline auto_round/utils/device.py]: cuda_xpu_devcies->gpu_devcies?

File: auto_round/utils/device.py [hunk 5]
```diff
@@ -458,34 +458,71 @@ def check_memory_availability(device, inputs, weight, org_seqlen, org_bs):
...
+
+            # memory * 2, because it contains grad tensor.
+            layer_memory_dict[layer_name] = {"param_memory": param_memory_gb * 2, "output_memory": output_memory_gb * 2}
 
     # Assuming bfloat16 or float32, input and output
     input_output_memory = 2 * sum(tensor.nbytes for tensor in input_ids) / 1024**3
+    if torch.xpu.is_available():
+        # https://github.com/intel/torch-xpu-ops/issues/2232
+        # sdpa on XPU takes more memory than expected. 2 from grad tensor
+        additional_memory = 9 * 2 + 1
```
Comment (yiliu30) [inline auto_round/utils/device.py]: It would be better to give a name for constant.

File: auto_round/utils/device.py [hunk 6]
```diff
@@ -628,8 +665,107 @@ def set_non_auto_device_map(
...
+
+    Strategy:
+    1. Sort layers by memory size (descending), preserve order for equal sizes
+    2. Assign largest N layers to higher-index devices (N = num_devices)
+    3. Remaining layers use memory availability + layer continuity scorings
+
+    Args:
+        layer_memory_dict (dict): Mapping of layer names to memory info (order preserved)
+        device_memory (dict): Available memory for each device (will be modified)
+        cuda_xpu_devices (list): List of device names (e.g., ["cuda:0", "cuda:1"])
```
Comment (yiliu30) [inline auto_round/utils/device.py]:
```suggestion
        gpus (list): List of device names (e.g., ["cuda:0", "cuda:1"])
```

File: auto_round/utils/device.py (hunk 2 above)
Comment (xin3he) [reply auto_round/utils/device.py]: I assume that non-quantized layers are small memory usage and could be ignored.
Do you mean the sdpa op by attention activations?

Comment (wenhuach21) [reply auto_round/utils/device.py]: 1 layer_config could set some Linear layers to be non-quantized, but at tuning phase, there should be on GPU, so we'd better not ignore their weights. And all the activations(layer output) should not be ignored if possible as they are not negligible.

2 yes, the naive attention is seqlen*seqlen, I am not familiar with sdpa/flash

File: auto_round/compressors/base.py [hunk 7]
```diff
@@ -2577,6 +2570,7 @@ def _quantize_block(
                 current_output = to_device(current_output, device)
 
                 output_q = self._get_current_q_output(block, input_ids, input_others, indices, device)
+                clear_memory()  # clean cached memory after getting output_q
```
Comment (wenhuach21) [inline auto_round/compressors/base.py]: do not add clear  memory in loop.  clear memory is a very time consuming op

File: auto_round/compressors/base.py [hunk 8]
```diff
@@ -2596,6 +2590,7 @@ def _quantize_block(
 
                 total_loss += loss.item() / num_elm
                 self._scale_loss_and_backward(scaler, loss)
+                clear_memory()  # clean cached memory after backward
```
Comment (wenhuach21) [inline auto_round/compressors/base.py]: same

File: auto_round/compressors/base.py [hunk 9]
```diff
@@ -313,6 +311,7 @@ def __init__(
         self.low_gpu_mem_usage = low_gpu_mem_usage
         self.seqlen = seqlen
         self.batch_size, self.gradient_accumulate_steps = batch_size, gradient_accumulate_steps
+        self.pick_samples = self.batch_size * self.gradient_accumulate_steps
```
Comment (wenhuach21) [inline auto_round/compressors/base.py]: as we have lost of members in this class. We'd better not to add new ones if possible

File: auto_round/compressors/base.py [hunk 10]
```diff
@@ -2488,9 +2483,8 @@ def _quantize_block(
         if q_input is not None:
             if input_ids is not q_input:
                 clear_memory(input_ids)
-            else:
-                clear_memory()
             input_ids = q_input
+        clear_memory()
```
Comment (wenhuach21) [inline auto_round/compressors/base.py]: same

File: auto_round/utils/device.py (hunk 2 above)
Comment (xin3he) [reply auto_round/utils/device.py]: from my knowledge, if flash attention is enabled, sequence_length² could be ignored. it could be the root cause of XPU.

File: auto_round/compressors/base.py [hunk 11]
```diff
@@ -2590,7 +2591,11 @@ def _quantize_block(
 
                 total_loss += loss.item() / num_elm
                 self._scale_loss_and_backward(scaler, loss)
-                clear_memory()  # clean cached memory after backward
+
+                # Temporary change for 70B model OOM issue on XPU
+                # TODO: Remove after https://github.com/intel/torch-xpu-ops/issues/2232 is fixed
+                if torch.xpu.is_available():
```
Comment (wenhuach21) [inline auto_round/compressors/base.py]: and self.low_gpu_mem_usage ?

File: auto_round/utils/device.py [hunk 12]
```diff
@@ -846,6 +846,16 @@ def set_auto_device_map_for_block_with_tuning(
...
 
+    # Ensure all remaining modules with params/buffers are moved to device_0
+    # This prevents mixed CPU/GPU execution within the same block
+    for name, module in block.named_modules():
+        if name not in names:  # This module wasn't assigned a device
+            # Check if module has any parameters or buffers
+            has_params = any(True for _ in module.parameters(recurse=False))
+            has_buffers = any(True for _ in module.buffers(recurse=False))
+            if has_params or has_buffers:
+                set_tuning_device_for_layer(block, name, device_0)
```
Comment (wenhuach21) [inline auto_round/utils/device.py]: to my knowledge, setting tuning device does not move the module to the device.  Better double check whether is correct, I guess the module need to move to the tuning device

**Comment by xin3he** [comment]
New changes:
- Observe the real memory usage on CUDA, and roughly estimate additional activation memory from SiLU and Norm by doubling layer outputs memory.
- pass output_device into set_auto_device_map_for_block_with_tuning
- add self.low_gpu_mem_usage as constrain

**Comment by xin3he** [comment]
### Cuda:0 memory used with no linear layers on cuda:0.
llama 8B W4A16
- about 6GB on CUDA

Qwen3 8B
- about  6GB on CUDA

Qwen3 32B
- about 10GB on CUDA

Now the estimation is matching the observation.

File: auto_round/compressors/base.py [hunk 13]
```diff
@@ -2750,7 +2750,6 @@ def _quantize_blocks(
                 modules = [get_module(model, n) for n in names]
                 m = WrapperMultiblock(modules)
 
-            m = m.to(device)
```
Comment (wenhuach21) [inline auto_round/compressors/base.py:2765]: Where is the code that moves the block to GPU when device_map=0? I haven’t seen it.
Also, please run the default command (auto-round --model xxx)for the 8B model to check whether the speed and accuracy are typically same with the main branch.

File: auto_round/utils/device.py [hunk 14]
```diff
@@ -628,8 +598,179 @@ def set_non_auto_device_map(
...
+    block_input_output_memory = 2 * sum(tensor.nbytes for tensor in input_ids) / 1024**3
+
+    # Roughly estimate additional memory for attention and other operations
+    additional_activation_memory = sum(info["output_memory"] for info in layer_memory_dict.values())
+    # 1GB considers norm weight, sdpa, reference_output, etc.
+    additional_memory = additional_activation_memory + 1  # GB
+    if torch.xpu.is_available():
+        # https://github.com/intel/torch-xpu-ops/issues/2232
+        # TODO: XPU takes more memory than expected. for llama 8B, it's about 12 GB
+        xpu_additional_memory = 12  # GB
```
Comment (yiliu30) [inline auto_round/utils/device.py:800]: Could you explain more about where the 12GB comes from? Is t related to the model? 
Would that be okay for 100B+ models?

Comment (xin3he) [reply auto_round/utils/device.py:800]: It's not okay, It's for 8b on XPU, and we should expect this issue is fixed in WW47

File: auto_round/utils/device.py [hunk 15]
```diff
@@ -415,6 +415,62 @@ def clear_memory(tensor=None):
...
+
+    Returns:
+        bool: True if memory was cleared, False otherwise.
+    """
+    # Check CUDA devices
+    if torch.cuda.is_available():
+        num_devices = torch.cuda.device_count()
+        for i in range(num_devices):
+            try:
+                total_memory = torch.cuda.get_device_properties(i).total_memory
```
Comment (wenhuach21) [inline auto_round/utils/device.py]: Would it be better to wrap some functions to make memory information retrieval device-agnostic

**Review by yiliu30 — APPROVED** [review]
